### PR TITLE
Support overriding encodeInput and decodeOutput

### PR DESCRIPTION
**Pull request by dekz**
## Description

Generated wrappers can now take an `encodeInput` and `decodeOutput` override allowing a different ABI encoder to be passed in.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: abi-gen/CHANGELOG.json [hunk 1]
```diff
@@ -1,4 +1,12 @@
 [
+    {
+        "version": "5.6.0",
+        "changes": [
+            {
+                "note": "Support overriding encodeInput and decodeOutput"
```
Comment (dorothy-zbornak) [inline abi-gen/CHANGELOG.json:6]: PR plx

File: base-contract/src/index.ts [hunk 2]
```diff
@@ -233,6 +240,18 @@ export class BaseContract {
...
         );
     }
+    // tslint:disable prefer-function-over-method
+    protected _isEncoderOverrides(arg: any): arg is EncoderOverrides {
+        return arg && (arg.encodeInput || arg.decodeOutput);
+    }
+    // tslint:disable prefer-function-over-method
+    protected _isEncoderRules(arg: any): arg is EncodingRules {
+        return arg && (arg as EncodingRules).shouldOptimize && arg;
+    }
```
Comment (dorothy-zbornak) [inline base-contract/src/index.ts:250]: the linter is right on this

File: abi-gen/templates/TypeScript/contract.handlebars [hunk 3]
```diff
@@ -381,9 +385,9 @@ export class {{contractName}}Contract extends BaseContract {
         txDefaults?: Partial<TxData>,
         logDecodeDependencies?: { [contractName: string]: ContractAbi },
         deployedBytecode: string | undefined = {{contractName}}Contract.deployedBytecode,
-        encodingRules?: EncodingRules,
+        encodingOpts?: EncodingRules | Partial<EncoderOverrides>,
```
Comment (dorothy-zbornak) [inline abi-gen/templates/TypeScript/contract.handlebars:388]: There's exactly one place in the world where we would ever pass in this final constructor arg, right? Why not just always have it be `EncoderOverrides`?

**Review by dorothy-zbornak — APPROVED** [review]

File: abi-gen/templates/TypeScript/contract.handlebars (hunk 3 above)
Comment (dekz) [reply abi-gen/templates/TypeScript/contract.handlebars]: Agreed, left it in for a/b tests. GET BE GONE

**Pull request merged**